nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/lifecycle/event/BulkLifeCycleChangeListener.java

```diff
@@ -159,17 +159,29 @@ protected void changeChildrenState(CoreSession session, String transition, Strin
             long pageSize = Long.parseLong(confService.getProperty(GET_CHILDREN_PAGE_SIZE_PROPERTY, "500"));
             // execute a first query to know total size
             String query = String.format("SELECT * FROM Document where ecm:parentId ='%s'", doc.getId());
-            DocumentModelList docs = session.query(query, null, pageSize, 0, true);
-            changeDocumentsState(session, transition, targetState, docs);
+            DocumentModelList documents = session.query(query, null, pageSize, 0, true);
+            changeDocumentsState(session, transition, targetState, documents);
+            session.save();
+            // commit the first page
+            TransactionHelper.commitOrRollbackTransaction();
+
             // loop on other children
-            long nbChildren = docs.totalSize();
-            for (long i = 1; i < nbChildren / pageSize; i++) {
-                docs = session.query(query, null, pageSize, pageSize * i, false);
-                changeDocumentsState(session, transition, targetState, docs);
+            long nbChildren = documents.totalSize();
+            for (long offset = pageSize; offset < nbChildren; offset += pageSize) {
+                // start a new transaction
+                TransactionHelper.startTransaction();
+                documents = session.query(query, null, pageSize, offset, false);
+                changeDocumentsState(session, transition, targetState, documents);
+                session.save();
+                TransactionHelper.commitOrRollbackTransaction();
```

Comment:
You can do this part more nicely with:

    TransactionHelper.runInTransaction(() -> {
        documents = session.query(query, null, pageSize, offset, false);
        changeDocumentsState(session, transition, targetState, documents);
        session.save();
    });